Solutions/Marain.TenantManagement.Abstractions/Marain/TenantManagement/Internal/TenantPropertyKeys.cs

```diff
@@ -0,0 +1,32 @@
+﻿// <copyright file="TenantPropertyKeys.cs" company="Endjin Limited">
+// Copyright (c) Endjin Limited. All rights reserved.
+// </copyright>
+
+namespace Marain.TenantManagement.Internal
+{
+    /// <summary>
+    /// Constants and helpers for well-known properties of tenants.
+    /// </summary>
+    public static class TenantPropertyKeys
+    {
+        /// <summary>
+        /// Key for storing the Service Manifest on a Service Tenant.
+        /// </summary>
+        public const string ServiceManifest = "Marain:ServiceManifest";
+
+        /// <summary>
+        /// Key for storing a tenant's enrollments.
+        /// </summary>
+        public const string Enrollments = "Marain:Enrollments";
+
+        /// <summary>
+        /// Builds the key to store the Id of a delegated service tenant that will be used by the specified service.
+        /// </summary>
+        /// <param name="serviceTenantId">
+        /// The Id of the Service Tenant representing the service that will be using the Delegated Tenant.
```

Comment:
So, from a diagnostics point of view, I agree service tenant name would be easier. If we do that, I think we are making it harder to ever allow service tenant names to change, but I don't particularly have a problem with this.

In terms of obliging any service that performs delegated work to look up the service tenant Id, I had been working on the basis that each service would have its corresponding service tenant Id set in its config. There's no particular issue with switching to using the name here either, just the point that it won't necessarily be possible for a service to have a hard coded name for its service tenant.

Reply:
Does this not mean you have to guarantee the the names are unique and suddenly you have all sorts of new issues.

Reply:
(and is why I initially resisted having a first class "name")

Reply:
matthew.adams:eyes:  00:11
What if we have two child tenants. One is yours and is called Foo. One is mine and is called Foo. This is not a good thing.

jonathan.george  08:43
I haven't got any tests that explicitly check it won't be a problem
But I don't think it would
I need service tenant names to be unique
Not so much client tenants

matthew.adams:eyes:  08:44
Well, we need to constrain that. If it is a requirement.

jonathan.george  08:44
Because?
Sorry
Misread that

matthew.adams:eyes:  08:45
Because we will forget. Or accidentally mess it up.

jonathan.george  08:45
I don't think it is a requirement

matthew.adams:eyes:  08:45
`I need service tenant names to be unique`
If you need that, then we have to enforce it.
I really really didn't want tenant name to become an identifier in any wya
And that's why I didn't give it a first class name.
Within a week of getting "Tenant Name" that is exactly what has happened. (edited) 
And others will assume it is unique in other circumstances (erroneously) in future too.
So I'm not happy.
The whole point was that IDs are guids and guaranteed unique.
And then everything else is decorative.
And you can look up a pretty name given a guid.

jonathan.george  08:48
Uniqueness of service tenant names is already enforced

matthew.adams:eyes:  08:48
But otherwise it is guids all the way down.
Is it?
Ok.

jonathan.george  08:48
In Marian.tenantmanagement, yes
Not if you go and mess around directly in the tenant provider

matthew.adams:eyes:  08:48
Ok.

jonathan.george  08:49
And the only reason I did that was to make service manifests vaguely readable

matthew.adams:eyes:  08:50
Oh, so I gave a better solution for that.
In the manifest have a table of "guid to pretty name" where the pretty name is local to the manifest
Then use the pretty names in the rest of the manifest
They are locally unique not globally unique.

jonathan.george  08:51
```{
  "$schema": "./ServiceManifestSchema.json",
  "contentType": "application/vnd.marain.tenancy.servicemanifests.servicemanifest",
  "serviceName": "Workflow v1",
  "dependsOnServiceNames": [ "Operations v1" ]
}
```
could do. The name is only used in that one array though.
so you may as well do something like
"dependendsOnServiceNames": [ { "name": "Operations v1", "id": "blah" } ]
but
it's slightly more than that
what we were shooting for is service manifest that you wouldn't have to modify as part of deployment

matthew.adams:eyes:  08:52
I'd still not use the name.

jonathan.george  08:52
if we do dependencies based on names, the manifest can remain the same across multiple deployments
if it's Id, we'd have to modify the manifests as we deployed the dependent services into a specific location

matthew.adams:eyes:  08:54
So name has become the unique identifier and the ID is just irrelevant.
You are basically just supplying a well known ID that is not a GUID.
My point, ultimately is that you've set up a competing ID and that just sounds like a world of pain.
I'd have a quick chat with Ian and (while you disagree!) explain my anxiety as if you agreed and see what he has to say :grinning:

jonathan.george  08:59
I get your point. I'd argue that the only point at which we are using the names is a) on service enrollment, to allow the manifests to stay the same cross-environment, and b) at the point of enrollment, to make it easier when people are using the CLI
(i.e. you can say enroll <tenant> <service name>
Ian has actually asked for something that takes is further down the route of reliance on tenant names - https://github.com/marain-dotnet/Marain.TenantManagement/pull/5#discussion_r404570807
Which I think stems from the fact that he's thinking about how he's going to deploy this stuff.
But I will talk it through with him again

matthew.adams:eyes:  09:01
I am totally fine with making Tenant IDs "explicit" as long as the unqiueness constraint is enforced.
But I don't like using names as IDs.
Bearing in mind that IDs contain the whole tenant path.
It is a massive can of worms.
I'd rather have a table in the management infrastructure that assigned "unique pretty names" to the GUIDs to do the lookup.

jonathan.george  09:03
Yes. Internally (up until Ian's comment) it's been IDs all the way

matthew.adams:eyes:  09:04
So you can assign arbitrary (but unique) vanity names when you create any tenant - and that becomes the TenantName which you can cheaply look-up the id from.
So enforce uniqueness globally (rather like Azure resource names)
That would be a simple change, and make it all symmetrical. (edited) 
Because I agree being able to use a simple name would be good - and it is basically just making "name" first class rather than an over-promoted general-purpose-label.

jonathan.george  09:06
It would. I don't think it's needed though, and I don't think we need to enforce unique names elsewhere.

matthew.adams:eyes:  09:06
You'll find other cases where you want it.

jonathan.george  09:06
If it bothers you that much, I'd rather just back off the idea of relying service names.

matthew.adams:eyes:  09:06
Such as looking up stuff in logging.

jonathan.george  09:06
hm

matthew.adams:eyes:  09:07
This is the thing - as soon as you want a "human readable" layer - you need a good place to do a quick lookup. (edited) 

jonathan.george  09:07
I've tried to be nice and include Ids and names in my log messages :slightly_smiling_face:

matthew.adams:eyes:  09:07
But your names are not unique - so you risk getting false positives.
So you need to enforce uniqueness to make it reliable.
:slightly_smiling_face:
There's no getting round it!

jonathan.george  09:08
So you're saying I should remove the names from my log messages.
:rolling_on_the_floor_laughing:
:joy:
Back to your original point
if we have two tenants
I don't see why we would want to stop both of them having a child with the same name
In fact, since "name" isn't used for anything, I think it would be odd to constrain them in that way.

matthew.adams:eyes:  09:10
But if you argue that you want enroll <tenant> <pretty service name>

jonathan.george  09:10
Right
But in Marain.TenantManagement we have essentially three different types of tenant (edited) 
Client
Service
and "Delegated" - the "on behalf of" tenants

matthew.adams:eyes:  09:11
surely you're going to want enroll MyRoyalFooness FrobulatorService rather than enroll de76ada8-e783-4909-aaaa-3431f386b6c5 FrobulatorService

jonathan.george  09:12
I haven't allowed that, because we can't rely on Client tenant names being unique :smile:

matthew.adams:eyes:  09:12
on exactly the same basis as you don't want enroll de76ada8-e783-4909-aaaa-3431f386b6c5 b4bc4ad1-857c-4fa4-8d2a-aced3bfb6dd0
Exactly - this is my point. If a human is writing that line, then you want human-readable names.
If it is being generated by tooling (which I assume it will be in the most common cases) then it doesn't matter.

jonathan.george  09:13
I wasn't expecting it to be a human writing that line
I was expecting it to be a developer

matthew.adams:eyes:  09:13
:smile:

jonathan.george  09:13
More to the point
I was expecting it to be me or Ian
:smile:
That's not entirely true
But I was expecting to wrap up some standard scripts "create a client tenant for a developer and enroll it in the standard services they need" so that some of this detail is hidden
but I really don't think we want to enforce uniqueness of client tenant names
if endjin bought services from two different companies using the same underlying infrastructure
it would be really odd if one of them turned round and said "yeah, sorry, we can't call you endjin in our system, you'll have to be endjin1"

matthew.adams:eyes:  09:15
Exactly.
I agree.

jonathan.george  09:16
Service tenants are a bit different though
For one thing, they all exist at the same level of the hierarchy, under a single parent
So it's relatively easy to do certain things with them - such as allow things at the edge, like the CLI - to refer to them by name (edited) 

matthew.adams:eyes:  09:18
OK - so would we create a pit of success if we could somehow identify them as "service tenants" and "client tenants" in the system

jonathan.george  09:18
Aha
So I toyed with the idea of adding a flag to each tenant to identify its type
Which I could then use for validation
At the moment they are various details about them that you could use to infer what type of tenant they are, if you so desired
for example, all service tenants a) have the same parent, and b) have a manifest

matthew.adams:eyes:  09:19
But if we made it explicit that would help a lot.
(Which can be a propertybag property. of course - it isn't something about "tenancy")

jonathan.george  09:20
I'm happy to do that. I came very close to doing it anyway, even though I didn't really need it.

matthew.adams:eyes:  09:20
(It is something about Marain Tenancy)

jonathan.george  09:20
But I will drop it in and add some extra validation off the back of it

matthew.adams:eyes:  09:21
I think that's good. Can you write the upshot of this conversation up in the channel.
Channel.
Bug
Work item
Issue
PR
THE WHATEVER :all-the-things:

jonathan.george  09:22
If by "write up", you mean "copy and paste this conversation".... :smile:

*Narrator*: And that is what happened.

Reply:
@jongeorge1 :
>each service would have its corresponding service tenant Id set in its config

When you say "its config" is that app settings?

So to be specific, `Marain.Workflow` would have something like this:

```
TenantIds:Marain.Tenancy:    1203f872308909ff8120348ad12
TenantIds:Marain.Operations: 5208c8ad121203f872308909ff8
```

in its App Settings (and equivalents in `local.settings.json` for local debugging)? This implies a new complication in the deployment process. If app settings are going to be set by ARM deployment (which is currently how it works) we now need the deployment process to somehow discover these values first.

This isn't impossible of course. I'm just checking I've understood.


@mwadams:
> Does this not mean you have to guarantee the the names are unique and suddenly you have all sorts of new issues.

I think we already have this problem, we've just removed the ability to enjoy the associated benefits. We are essentially still bound to the name because the IDs are not fixed—every tenant gets a randomly assigned ID when created, meaning that no two marain instances will have the same ID for any particular service. So how do I find out which ID to use? Well I just look for the service with the matching name...

So the only thing that identifies "Workflow v1" once and for all is its name. Yes, in one particular tenant it will have one particular ID, but to discover what the ID is, I have to look at every service tenant and find the one with the matching name.

What we've got right now is a bit like it COM were modified so that when you installed a new COM component, you didn't get to say what the CLSID is, and instead you are assigned a brand new GUID that's different on every machine. At this point the ProgId (the textual name for the component, the means by which scripting languages generally identified the class they want) becomes the only stable identifier if you want script to work on more than one environment, and even within a single environment, the textual name (ProgId) would be the means by which you find the CLSID you need. CLSIDs would be essentially a pointless extra layer in such a system. Their value in reality depends on the fact that individual services get to define once and for all what their CLSID is.

If we modified things so that we could specify our own ID when creating tenants, then we could make it look more like COM, where these IDs are fixed once and for all, and then they'd provide a useful global identity mechanism. If `Marain.Tenancy` could be counted on always to have the ID `1203f872308909ff8120348ad12` then that suddenly becomes a whole lot more useful because you're no longer having to go by the display name first, but currently this is not an option. The tenant ID is determined for you by the tenancy service.

So we have all the readability/debuggability downsides of alphanumeric GUIDs but combined with all the uniqness problems of text names because we're still ultimately dependent on those text strings to define identity.

My reason for suggesting using text names is that we already implicitly depend on the name as the only stable identity so we may as well enjoy the benefits. But if we're up for making the necessary changes to enable "bring your own GUI" on tenant creation then I'm perfectly happy to go down the "alphanumeric IDs are the one true manifestation of identity" path because it solves a bunch of unique naming problems as Matthew said, and also because I still have a soft spot for COM (although that might be Stockholm syndrome).

Reply:
I know there has been some further discussion, but "Stable ID" would be preferably to me than "Unique name" (as we would ideally have one and only one ID)

So why not make the IDs stable, rather like RootTenantID is stable?

Reply:
Yes, I'd be happy with stable IDs. So we'd need an overload of or optional argument for `ITenantProvider.CreateChildTenantAsync` enabling you to bring your own GUID, instead of having it append one generated by `NewGuid()`.

Assuming the parent tenant ID is stable (which, if it's the root tenant, it will be) then the newly created tenabt's ID will also be stable.

Reply:
And I think I might call it something like `CreateWellKnownTenantAsync()`

We could then document the constraints of that (that it troubles to check for uniqueness within the parent tenant etc.)

Reply:
Ok. I'm going to merge this PR anyway, but will create new issues in Corvus.Tenancy, Marain.Tenancy and Marain.TenantManagement to address this change.